W5500.device.nut

```diff
@@ -0,0 +1,2256 @@
+    // Parameters:
+    //      spi - configured spi, W5500 supports spi mode 0 or 3
+    //      cs(optional) - configured chip select pin
+    //      reset(optional) - configured reset pin
+    // ***************************************************************************
+    constructor(spi, cs = null, resetPin = null) {
+        _spi = spi;
+        _cs = cs;
+        _resetPin = resetPin;
+        _connections = {};
+        _availableSockets = [];
+
+        // Check the pins
+        if (resetPin) _resetPin.configure(DIGITAL_OUT, 1);
+
+        local imp005 = ("spi0" in hardware);
+        if (_cs != null) {
+            _cs.configure(DIGITAL_OUT, 1);
+        } else if (!imp005) {
+            throw "You must pass in a chip select pin."
+        }
+
+        // Reset the hardware
+        reset();
+
+        // server.log("Wiznet chip version: " + getChipVersion());
+
+    }
+
+    // ***************************************************************************
+    // init(cb)
+    // Returns: nothing
+    // Parameters:
+    //        cb                        - callback function
+    //        remaining(optional)       - for first time Initialise == null
+    //                                  - subseuqent goes to to else
+    //        closes all sockets and fires the callback when complete
+    // ***************************************************************************
+    function init(cb, remaining = null) {
+
+        if (remaining == null) {
+
+            // Initialise the number of sockets
+            _maxNoOfSockets = getTotalSupportedSockets();
+
+            // The first time through, check what needs to be dealt with
+            remaining = [];
+            for (local socket = 0; socket < _maxNoOfSockets; socket++) {
+                local status = getSocketStatus(socket);
+                if (status == W5500_SOCKET_STATUS_ESTABLISHED) {
+                    sendSocketCommand(socket, W5500_SOCKET_DISCONNECT);
+                    remaining.push({ socket = socket, status = status });
+                } else if (status != W5500_SOCKET_STATUS_CLOSED) {
+                    remaining.push({ socket = socket, status = status });
+                }
+            }
+
+        } else {
+
+            // The second+ time through
+            local oldremaining = remaining;
+            remaining = [];
+            foreach (prev in oldremaining) {
+
+                local socket = prev.socket;
+                local prevstatus = prev.status;
+                local newstatus = getSocketStatus(socket);
+
+                if (newstatus == W5500_SOCKET_STATUS_CLOSED) {
+                    sendSocketCommand(socket, W5500_SOCKET_CLOSE);
+                } else {
+                    remaining.push({ socket = socket, status = newstatus });
+                }
+
+            }
+
+        }
+
+        if (remaining.len() == 0) {
+
+            // Set the defaults
+            _availableSockets = setNumberOfAvailableSockets(4);
```

Comment:
It turns out that changing from 4 to 8 reduces the write buffer size to the point that DHCP no longer works. It took me a while to work this out :(